### PR TITLE
CBSD VM discovery

### PR DESCRIPTION
**Pull request by VVelox**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [X] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Still working on the extend portion, but this is all good.

Need to look into adding graphing and the like for max ram, ram usage, cpu usage, and current number of procs.

### CODE REVIEW

**Comment by VVelox** [comment]
Worth nothing vmwVmVMID is always set to zero as CBSD does not produce a numeric ID really. Well there is, but this is the process ID of the main bhyve proc running it and it will change between VM restarts because of that.

File: includes/discovery/cbsd-vminfo.inc.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    $cbsd_keys=array_keys($cbsd[data]);
+
+    foreach ($cbsd_keys as $vm) {
+        $vmwVmDisplayName = $cbsd{data}{$vm}{jname};
+        $vmwVmGuestOS     = $cbsd{data}{$vm}{vm_os_type};
+        $vmwVmMemSize     = $cbsd{data}{$vm}{vm_ram};
+        $vmwVmState       = $cbsd{data}{$vm}{status};
+        $vmwVmCpus        = $cbsd{data}{$vm}{vm_cpus};
```
Comment (PipoCanaja) [inline includes/discovery/cbsd-vminfo.inc.php]: You need to write all hash as $myHash{'key'} :)

**Comment by PipoCanaja** [comment]
You still have quite a few keys to updates, and you also use '{' instead of '[' to enclose the keys. Could you change that to a more standard syntax ?

File: includes/discovery/cbsd-vminfo.inc.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    /*
+     * Fetch information about Virtual Machines.
+     */
+    try {
+        $cbsd=json_app_get($device, 'cbsd-bls', 1);
+    } catch (JsonAppException $e) {
+        # This is just empty and exists to catch it not existing.
+    }
+
+    $cbsd_keys=array_keys($cbsd[data]);
```
Comment (PipoCanaja) [inline includes/discovery/cbsd-vminfo.inc.php]: another one :)

Comment (VVelox) [reply includes/discovery/cbsd-vminfo.inc.php]: Doh! You got to this first.

**Comment by VVelox** [comment]
@PipoCanaja Howdy! Sorry about the slow reply on this. Thanks for noticing the slightly out of date comment chunk there. :)

Also I think this is alll good now. Was just giving this another look over and not seeing anything that does not meet the current standards off hand.

**Comment by murrant** [comment]
This requires an snmp extend script? You probably need some docs for it.

**Comment by VVelox** [comment]
@murrant Actually going to be shelving this for a bit. I really want to put more time into actually making pulling this info actually useful for something.

Currently we can for VMware and libvirt and it is not really used. This would just add a third we don't really do anything with this with.

Need to begin actually implementing actually graphing and the like based on this.

**Pull request closed without merging**